src/uct/ib/base/ib_iface.h

```diff
@@ -357,4 +359,66 @@ void uct_ib_iface_set_max_iov(uct_ib_iface_t *iface, size_t max_iov)
 }
 
 
+static UCS_F_ALWAYS_INLINE
+void uct_ib_iface_fill_ah_attr_raw(
+        uct_ib_iface_t *iface, uint16_t lid,
+        union ibv_gid *gid, uint8_t *path_bits_p,
+        struct ibv_ah_attr *ah_attr)
+{
+    uint8_t path_bits = (path_bits_p != NULL) ? *path_bits_p : iface->path_bits[0];
+
+    memset(ah_attr, 0, sizeof(*ah_attr));
+
+    ah_attr->sl                = iface->config.sl;
+    ah_attr->src_path_bits     = path_bits;
+    ah_attr->dlid              = lid | path_bits;
+    ah_attr->port_num          = iface->config.port_num;
+    ah_attr->grh.traffic_class = iface->config.traffic_class;
+
+    if ((gid != NULL) &&
+        ((iface->addr_type == UCT_IB_ADDRESS_TYPE_ETH) ||
+         (iface->addr_type == UCT_IB_ADDRESS_TYPE_GLOBAL) ||
+         (iface->gid.global.subnet_prefix != gid->global.subnet_prefix))) {
+        ah_attr->is_global = 1;
+        ah_attr->grh.sgid_index = iface->config.gid_index;
+        ah_attr->grh.dgid = *gid;
+    } else {
+        ah_attr->is_global = 0;
+    }
+}
+
+static UCS_F_ALWAYS_INLINE
+void uct_ib_iface_fill_ah_attr_from_addr(
+        uct_ib_iface_t *iface, const uct_ib_address_t *ib_addr,
+        uint8_t path_bits, struct ibv_ah_attr *ah_attr)
+{
+    union ibv_gid *gid_p = NULL;
+    union ibv_gid  gid;
+    uint8_t        is_global;
+    uint16_t       lid;
+
+    uct_ib_address_unpack(ib_addr, &lid, &is_global, &gid);
+
+    /* Create a global address only if we cannot reach the destination using local
+     * address. It means either an Ethernet address, or IB address on a different subnet.
+     */
+    if (is_global &&
+        ((iface->addr_type == UCT_IB_ADDRESS_TYPE_ETH) ||
```

Comment:
maybe check for is_global is enough?

Reply:
nop, for *LOCAL addressing and same subnet (gid is defined there) lid-based adressing should be used

Reply:
need to remove this check since it's already inside uct_ib_iface_fill_ah_attr_raw(). 
and below, need to:
`uct_ib_iface_fill_ah_attr_raw(iface, lid, is_global ? &gid: NULL, &path_bits, ah_attr);`

Reply:
will fix